src/EntryPoint.sol

```diff
@@ -816,37 +797,54 @@ contract EntryPoint is
+            TokenTransferLib.balanceOf(u.paymentToken, u.paymentRecipient), paymentAmount
         );
-        address eoa = u.eoa;
-        address payer = Math.coalesce(u.payer, eoa);
-        if (paymentAmount > u.paymentMaxAmount) {
-            revert PaymentError();
-        }
-        bytes calldata paymentSignature = u.paymentSignature;
+
+        address payer = Math.coalesce(u.payer, u.eoa);
+
+        // Call the pay function on the delegation contract
+        // Equivalent Solidity code:
+        // IDelegation(payer).pay(paymentAmount, keyHash, abi.encode(u));
+        // Gas Savings:
+        // Saves ~2k gas for normal use cases, by avoiding abi.encode and solidity external call overhead
         assembly ("memory-safe") {
-            let m := mload(0x40) // Cache the free memory pointer.
-            mstore(m, 0xce835432) // `compensate(address,address,uint256,address,bytes32,bytes32,bytes)`.
-            mstore(add(m, 0x20), shr(96, shl(96, paymentToken)))
-            mstore(add(m, 0x40), address())
-            mstore(add(m, 0x60), paymentAmount)
-            mstore(add(m, 0x80), shr(96, shl(96, eoa)))
-            mstore(add(m, 0xa0), keyHash)
-            mstore(add(m, 0xc0), digest)
-            mstore(add(m, 0xe0), 0xe0)
-            mstore(add(m, 0x100), paymentSignature.length)
-            calldatacopy(add(m, 0x120), paymentSignature.offset, paymentSignature.length)
-            pop(
-                call(gas(), payer, 0, add(m, 0x1c), add(0x104, paymentSignature.length), 0x00, 0x00)
-            )
+            let m := mload(0x40) // Load the free memory pointer
+            mstore(m, 0xf81d87a7) // `pay(uint256,bytes32,bytes32,bytes)`
+            mstore(add(m, 0x20), paymentAmount) // Add payment amount as first param
+            mstore(add(m, 0x40), keyHash) // Add keyHash as second param
+            mstore(add(m, 0x60), digest) // Add digest as third param
+            mstore(add(m, 0x80), 0x80) // Add offset of encoded UserOp as third param
+
+            let encodedSize := sub(calldatasize(), u)
+
+            mstore(add(m, 0xa0), add(encodedSize, 0x20)) // Store length of encoded UserOp at offset.
+            mstore(add(m, 0xc0), 0x20) // Offset at which the UserOp struct starts in encoded UserOp.
+
+            // Copy the userOp data to memory
+            calldatacopy(add(m, 0xe0), u, encodedSize)
+
+            // TODO: If pay reverts, we now send a revert back instead of ignoring. This is a breaking change, add to changeset.
+            if iszero(
+                call(
+                    gas(), // gas
+                    payer, // address
+                    0, // value
+                    add(m, 0x1c), // input memory offset
+                    add(0xc4, encodedSize), // input size
+                    0x00, // output memory offset
+                    0x20 // output size
+                )
+            ) { revert(0x00, 0x20) }
```

Comment:
Maybe add a comment on rationale on reverting. Like bubble up the revert for debugging purposes.